Create Kpi page

lehecht opened the pull request:
- Write bash script to count visits and actions
- Write API to save number of actions/visits
- Determine and save number of (unique) user and storage usage
- Show kpi page in admin area

Resolves #1

lehecht committed:
Add migrations

lehecht committed:
Add paths of migration, config and assets

lehecht committed:
Add scheduler for unique user counting task

lehecht committed:
Create task to count unique user

lehecht committed:
Create daily task to count user

lehecht committed:
Create config file

lehecht committed:
Add tests for user counting

lehecht committed:
Create task to determine storage usage

lehecht committed:
Add test

lehecht committed:
Add scheduler to determine storage usage

lehecht committed:
Add scheduler to count user daily

lehecht committed:
Remove superfluous kernel

lehecht committed:
Add router to provider

lehecht committed:
Add route for actions/visits

lehecht committed:
Save actions/visits

lehecht committed:
Add controller for action/visits

lehecht committed:
Add middleware to route

lehecht committed:
Add middleware

lehecht committed:
Seperate action/visit route from middleware

Because this route is not used by a user, the CSRF-Token is missing.
Web-middleware checks for a valid CSFR-Token,
which is why this route does not use this middleware.

lehecht committed:
Add comment

lehecht committed:
Add routes to get counts

lehecht committed:
Add controller for User and Storage

lehecht committed:
Add tests

lehecht committed:
Fix wrong date

lehecht committed:
Fix tests

lehecht committed:
Change save date from first to last day of month

lehecht committed:
Update tests

lehecht committed:
Check for empty array

lehecht committed:
Add tests

lehecht committed:
Add view to display kpis

lehecht committed:
Remove unused controller

lehecht committed:
Remove unused routes

lehecht committed:
Add get methods

lehecht committed:
Load views in servic provider

lehecht committed:
Add request validation

lehecht committed:
Add response for missing token

lehecht committed:
Change route path

lehecht committed:
Add tests

lehecht committed:
Handle invalid month index number

If month index is invalid,
use index of current month instead.

lehecht committed:
Fix wrong user count by using sum function

lehecht committed:
Fix month options

Remove current month from options,
because user are still counted until end of month.

lehecht committed:
Add tests for actions/visits

lehecht committed:
Add kpi button to admin menu

lehecht committed:
Add storage test

lehecht committed:
Add test for user count

lehecht committed:
Fix wrong dates in user count

lehecht committed:
Add script to count server requests

lehecht committed:
Add array size check before value check

lehecht committed:
Remove unused routes file

lehecht committed:
Make url parameter optional of kpi admin page

lehecht committed:
Fix inactive button by using wildcard

lehecht committed:
Make date unique

lehecht committed:
Remove string length from rule set

lehecht committed:
Fix UniqueConstraintViolationException

Fix tests by splitting them up.

lehecht committed:
Move test method

lehecht committed:
Fix wrong storage usage computation

lehecht committed:
Add token to config

lehecht committed:
Use token from config file

lehecht committed:
Remove token from script

lehecht committed:
Align buttons with panels

lehecht committed:
Add sixth month to overview

mzur commented on src/Console/Commands/CountUniqueUser.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        $fromDate = Carbon::now()->subMonth()->startOfMonth()->toDateString();
+        $tillDate = Carbon::now()->subMonth()->endOfMonth()->toDateString();
+        return User::whereBetween('login_at', [$fromDate, $tillDate])->get()->count();
```
You can use this as a more efficient query:
```suggestion
        return User::whereBetween('login_at', [$fromDate, $tillDate])->count();
```
This counts the rows directly in the database which is much faster.

mzur commented on src/Console/Commands/CountUser.php:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @var string
+     */
+    protected $description = 'Counts active user for each day';
```
```suggestion
    protected $description = 'Counts active users of the previous day';
```

mzur commented on src/Console/Commands/CountUniqueUser.php:
```diff
@@ -0,0 +1,48 @@
+
+        DB::transaction(function () use ($nbrUser, $date) {
+            DB::table('kpis_unique_users')->insert(['date' => $date, 'value' => $nbrUser]);
+        });
```
There is no need to wrap a single query in a transaction block. It only makes sense for multiple queries.

mzur commented on src/Console/Commands/CountUser.php:
```diff
@@ -0,0 +1,46 @@
+
+        DB::transaction(function () use ($nbrUser) {
+            DB::table('kpis_users')->insert(['date' => Carbon::yesterday()->toDateString(), 'value' => $nbrUser]);
+        });
```
Same than above: No need for a transaction block.

mzur commented on src/Console/Commands/CountUser.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $today = Carbon::yesterday()->toDateString();
+
+        return User::where('login_at', '=', $today)->get()->count();
```
```suggestion
        return User::where('login_at', '=', $today)->count();
```

mzur commented on src/Console/Commands/DetermineStorageUsage.php:
```diff
@@ -0,0 +1,58 @@
+
+        DB::transaction(function () use ($size, $date) {
+            DB::table('kpis_storage_usage')->insert(['date' => $date, 'value' => $size]);
+        });
```
No need for a transaction block.

mzur commented on src/Console/Commands/DetermineStorageUsage.php:
```diff
@@ -0,0 +1,58 @@
+        $videoStorageUsage = Video::all()->reduce(function($res, $video){
+            $res += $video->getSizeAttribute();
+            return $res;
+        }, 0);
```
This will blow up in the production database with millions of files because with `get()` you attempt to get *all* of them at once.

If you have to do a loop with lots of data you can do it like this:
```php
$size = 0;
foreach (Image::lazyById() as $image) {
   $size += $image->size;
}
```
This will fetch the images in chunks from the database and loop over each one.

**But** there is a better way suggested in my [old comment](https://github.com/biigle/kpis/issues/1#issuecomment-1980562477). This is much faster and more efficient because everything is done in the database:

```php
$imageStorageUsage = Image::sum(DB::raw("(attrs->>'size')::bigint"));
$videoStorageUsage = Video::sum(DB::raw("(attrs->>'size')::bigint"));
```

I'm happy to explain the query if you are interested.

mzur commented on src/Database/migrations/2024_03_06_091955_create_kpis_unique_users.php:
I usually create all tables of a new feature in a single migration. Could you please merge everything into one file (e.g. `create_kpis_tables`)?

mzur commented on src/Http/Requests/StoreRequest.php:
```diff
@@ -0,0 +1,52 @@
+    public function rules()
+    {
+        return [
+            'value' => 'required|string',
```
Please have a look at the comment to `countRequests.sh` below for how a JSON request can be sent and validated. Then there will be no need for additional validation in `withValidator()` any more.

mzur commented on src/Http/Controllers/RequestController.php:
```diff
@@ -0,0 +1,23 @@
+     *
+     * @param StoreRequest $request containing the json object
+     *
+    */
```
Please add API documentation like [this](https://github.com/biigle/largo/blob/9592f32dae61ef8141477a3f1cfdcab2c33db0e3/src/Http/Controllers/Api/Volumes/LargoController.php#L13-L48). This also requires that you uncomment the [`apidoc` line](https://github.com/biigle/kpis/pull/4/files#diff-d1a2fad8ed7d2cb1d5f1e65323001a06e2a53193a06782feb0e4bc87dfc34c5dR49) in the service provider.

mzur commented on src/Http/Controllers/RequestController.php:
```diff
@@ -0,0 +1,23 @@
+    {
+        $res = json_decode($request->get('value'), true);
+        Requests::save($res['actions'], $res['visits']);
+        return;
```
No need for that.
```suggestion
```

mzur commented on tests/Http/Controllers/RequestControllerTest.php:
```diff
@@ -0,0 +1,54 @@
+
+    public function testStore()
+    {
+        $this->withHeader('Authorization', $this->testToken)
```
Shouldn't this be:
```suggestion
        $this->withHeader('Authorization', "Bearer {$this->testToken}")
```

mzur commented on tests/Http/Controllers/RequestControllerTest.php:
You should also check something like:
```php
 $this->withHeader('Authorization', $this->testToken)
        ->postJson($this->route, ['visits' => 'abc', 'actions' => 'def'])
        ->assertStatus(422);
```

mzur replied:
This already assumes that you have updated the form request to handle the actual JSON payload (see comment to `countRequests.sh` below).

mzur commented on src/Http/Controllers/Views/KpiController.php:
```diff
@@ -0,0 +1,51 @@
+    public function show($idx = 6)
+    {
+        // If index is not in range, set it on index of last month
+        $idx = $idx == 0 || $idx > 6 ? 6 : $idx;
```
What about `$idx = -1`? :wink:
```suggestion
        $idx = min(6, max(0, $idx));
```

lehecht replied:
This case is catched [here](https://github.com/biigle/kpis/blob/564c45a02a0a31571fe27b149600fd1d84ff2315/src/Http/web.php#L9).

mzur replied:
Ok. Then there will be a 404 response for -1 and you should probably also return a 404 for `$idx > 6`.

mzur commented on src/Http/Controllers/Views/KpiController.php:
```diff
@@ -0,0 +1,51 @@
+        $monthOverview = [];
+        for($i = 6;$i >= 1;$i--) {
+            $monthName = Carbon::now()->subMonths($i)->format('F');
+            $monthOverview[] = substr($monthName, 0, 3);
```
There is also:
```suggestion
            $monthOverview[]  = Carbon::now()->subMonths($i)->format('M');
```

mzur commented on tests/Http/Controllers/RequestControllerTest.php:
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Biigle\Modules\Kpis\Tests\Http\Controllers;
```
```suggestion
namespace Biigle\Tests\Modules\Kpis\Http\Controllers;
```

mzur commented on tests/Http/Controllers/Views/KpiControllerTest.php:
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Biigle\Modules\Kpis\Tests\Http\Controllers\Views;
```
```suggestion
namespace Biigle\Tests\Modules\Kpis\Http\Controllers\Views;
```

mzur commented on tests/RequestsTest.php:
```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Biigle\Modules\Tests;
```
```suggestion
namespace Biigle\Tests\Modules\Kpis;
```

mzur commented on tests/StorageTest.php:
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace Biigle\Modules\Kpis;
```
```suggestion
namespace Biigle\Tests\Modules\Kpis;
```

mzur commented on tests/UserTest.php:
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Biigle\Modules\Kpis;
```
```suggestion
namespace Biigle\Tests\Modules\Kpis;
```

mzur commented on src/Http/api.php:
```diff
@@ -0,0 +1,5 @@
+
+$router->post('api/v1/kpis', [
+    'uses' => 'RequestController@store'
+])->middleware('auth.kpis');
```
This does not work (for me) because the middleware alias is never defined. Instead you can directly specify the middleware class:

```suggestion
use Biigle\Modules\Kpis\Http\Middleware\EnsureTokenIsValid;

$router->post('api/v1/kpis', [
    'uses' => 'RequestController@store'
])->middleware(EnsureTokenIsValid::class);
```

mzur commented on src/Http/Middleware/EnsureTokenIsValid.php:
```diff
@@ -0,0 +1,28 @@
+            return response('Unauthorized', 401);
+        }
+
+        if ($request->header('authorization') !== config('kpis.token')) {
```
This should also handle the case where the header value has a `Bearer ` prefix (see comment in test case).

mzur commented on src/KpisServiceProvider.php:
```diff
@@ -42,6 +53,30 @@ public function boot(Modules $modules, Router $router)
+                $schedule = app(Schedule::class);
+
+                $schedule->command(CountUniqueUser::class)
+                    ->monthlyOn(1, '0:0')
```
This is equivalent to:
```suggestion
                    ->monthly()
```

mzur commented on src/KpisServiceProvider.php:
```diff
@@ -42,6 +53,30 @@ public function boot(Modules $modules, Router $router)
+                    ->onOneServer();
+
+                $schedule->command(DetermineStorageUsage::class)
+                    ->monthlyOn(1, '0:0')
```
```suggestion
                    ->monthly()
```

mzur commented on src/Requests.php:
```diff
@@ -0,0 +1,33 @@
+
+    public static function getActions($year, $month)
+    {
+        $date = Carbon::createFromDate($year, $month, 1)->endOfMonth()->toDateString();
+        $res =  count(DB::table('kpis_actions')->where('date', '=', $date)->pluck('value')) != 0 ?
+        DB::table('kpis_actions')->where('date', '=', $date)->pluck('value')[0] : 0;
```
Always try to do as much work as possible directly in the DB. Also the query seems wrong as it returns only the actions of the last day of the month and not the whole month:

```suggestion
        $start = Carbon::createFromDate($year, $month, 1)->toDateString();
        $end = Carbon::createFromDate($year, $month, 1)->endOfMonth()->toDateString();
        $res = DB::table('kpis_actions')->whereBetween('date', [$start, $end])->sum('value');
```

mzur replied:
Please also add a test case that checks if `$end` is included or not. I'm not sure how `whereBetween()` behaves with these dates.

lehecht replied:
Ah sorry. I read 'daily' in the issue but had monthly in my mind somehow.
But my [tests](https://github.com/biigle/kpis/blob/564c45a02a0a31571fe27b149600fd1d84ff2315/tests/RequestsTest.php#L45-L64) already use the last day of the month to check the values. Or did I misunderstand you?

mzur replied:
No, the tests should be fine then!

mzur commented on src/Requests.php:
```diff
@@ -0,0 +1,33 @@
+    {
+        $date = Carbon::createFromDate($year, $month, 1)->endOfMonth()->toDateString();
+        $res =  count(DB::table('kpis_visits')->where('date', '=', $date)->pluck('value')) != 0 ?
+        DB::table('kpis_visits')->where('date', '=', $date)->pluck('value')[0] : 0;
```
Same than above.

```suggestion
        $start = Carbon::createFromDate($year, $month, 1)->toDateString();
        $end = Carbon::createFromDate($year, $month, 1)->endOfMonth()->toDateString();
        $res = DB::table('kpis_visits')->whereBetween('date', [$start, $end])->sum('value');
```

mzur commented on src/Storage.php:
```diff
@@ -0,0 +1,17 @@
+    {
+        $date = Carbon::createFromDate($year, $month, 1)->endOfMonth()->toDateString();
+        $res = count(DB::table('kpis_storage_usage')->where('date', '=', $date)->pluck('value')) != 0 ?
+        DB::table('kpis_storage_usage')->where('date', '=', $date)->pluck('value')[0] : 0;
```
While it's always just one row, `sum()` is convenient and returns 0 if no row exists.
```suggestion
        $res = DB::table('kpis_storage_usage')->where('date', '=', $date)->sum('value');
```

mzur commented on src/User.php:
```diff
@@ -0,0 +1,23 @@
+    public static function getUser($year, $month)
+    {
+        $first = Carbon::createFromDate($year, $month, 1);
+        $last = $first->copy()->endOfMonth();
+        return DB::table('kpis_users')->whereBetween('date', [$first->toDateString(), $last->toDateString()])->sum('value');
```
That's a good query :wink: Please confirm if `$last` is included or not in a test case (see above).

mzur commented on src/User.php:
```diff
@@ -0,0 +1,23 @@
+    {
+        $date = Carbon::createFromDate($year, $month, 1)->endOfMonth()->toDateString();
+        $res =  count(DB::table('kpis_unique_users')->where('date', '=', $date)->pluck('value')) != 0 ?
+        DB::table('kpis_unique_users')->where('date', '=', $date)->pluck('value')[0] : 0;
```
```suggestion
        $res = DB::table('kpis_unique_users')->where('date', '=', $date)->sum('value');
```

mzur commented on src/config/kpis.php:
```diff
@@ -0,0 +1,5 @@
+<?php
+
+return [
+    'token' => env('KPI_TOKEN'),
```
Please add a comment that explains what the token is for.

mzur replied:
Also, my (undocumented) env variable naming convention is to use `KPIS_TOKEN` for a config variable `kpis.token`.

```suggestion
    'token' => env('KPIS_TOKEN'),
```

mzur commented on src/resources/scripts/countRequests.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+
```
Please add a comment that explains how the script should be called to process a logfile.

mzur commented on src/resources/scripts/countRequests.sh:
```diff
@@ -0,0 +1,28 @@
+    bot="[bB]ot|OpenVAS-VT";
+    if [[ "$line" =~ $bot ]]; then
+        continue;
+    fi
```
Isn't this done implicitly with the next `if`?

mzur commented on src/resources/scripts/countRequests.sh:
```diff
@@ -0,0 +1,28 @@
+
+res="{\"visits\": $nbrVisits, \"actions\": $nbrActions}"
+
+curl -X POST -H "Authorization: Bearer <token>" -F "value=$res" https://biigle.de/api/v1/kpis -v
```
This is how you can send a JSON payload that won't be treated as plain string by Laravel:

```suggestion
curl -X POST \
   -H "Authorization: Bearer <token>" \
   -H "Content-Type: application/json" \
   -d "{\"visits\": $nbrVisits, \"actions\": $nbrActions}" \
   https://biigle.de/api/v1/kpis
```

The request above will have the `visits` and `actions` parameters, so the request validation could be done like this:

```php
['actions' => 'required|integer', 'visits' => 'required|integer']
```

And the values can be accessed like this:

```php
$visits = $request->input('visits');
```

mzur commented on src/resources/scripts/countRequests.sh:
```diff
@@ -0,0 +1,28 @@
+    browser="Chrome|Edg|Safari|OPR|Firefox";
+    if [[ "$line" =~ $browser ]]; then
+    # filter http code (2xx) by using regex of ' <http code> XY ' 
+      code="[[:space:]]2[[:digit:]]{2}[[:space:]][[:digit:]]+[[:space:]]"
```
It may be safer to match the code and HTTP verb by their position as determined by the space separators. Example:
```suggestion
      code=$(echo "$line" | cut -d ' ' -f 9)
      verb=$(echo "$line" | cut -d ' ' -f 6 | cut -c 2-)
```

While the code regex looks reasonably specific, the verb could easily match an URL (e.g. `https://biigle.de/POSTER`).

lehecht replied:
When I apply your changes and try to run the script, it takes up to 10 minutes to terminate. Is my machine just slow or does the time not matter in this case?

mzur replied:
If the script was faster before on your machine, it's slower now :smiley: Faster is better. I'll have a look.

mzur replied:
You are right, it's quite horrible with the change proposed above. This works quite fast:

```bash
#!/bin/bash
#
# Usage: ./countRequests.sh logfile.log.gz

TOKEN=""
BASE_URL="https://biigle.de"

compute_visits() {
    zgrep -F -e Chrome -e Firefox -e Safari -e OPR -e Edg $1 \
        | cut -d ' ' -f 6,9 \
        | cut -c 2-6 \
        | grep -F "GET 2" \
        | wc -l
}

compute_actions() {
    zgrep -F -e Chrome -e Firefox -e Safari -e OPR -e Edg $1 \
        | grep -F -v heartbeat \
        | cut -d ' ' -f 6 \
        | cut -c 2- \
        | grep -F -e PUT -e POST -e DELETE \
        | wc -l
}

nbrVisits=$(compute_visits $1)
nbrActions=$(compute_actions $1)

curl \
    -X POST \
    -H "Authorization: Bearer $TOKEN" \
    -H "Content-Type: application/json" \
    -d "{\"visits\": $nbrVisits, \"actions\": $nbrActions}" \
    ${BASE_URL}/api/v1/kpis

```

It also works with the gzipped logfiles and ignores the automatic "heartbeat" action.

mzur replied:
Thinking about it, the script above does not report the numbers described in #1. The visits should only count the dashboard/home page. The actions should also include automated API requests (via token authentication). Here is an updated script:

```bash
#!/bin/bash
#
# Usage: ./countRequests.sh logfile.log.gz

TOKEN=""
BASE_URL="https://biigle.de"

compute_visits() {
    zgrep -F -e Chrome -e Firefox -e Safari -e OPR -e Edg $1 \
        | cut -d ' ' -f 6,7,9 \
        | cut -c 2-8 \
        | grep -F "GET / 2" \
        | wc -l
}


compute_actions() {
    # The "@" will appear in API requests with token authentication.
    zgrep -F -e PUT -e POST -e DELETE $1 \
        | grep -F -e Chrome -e Firefox -e Safari -e OPR -e Edg -e "@" \
        | grep -F -v heartbeat \
        | wc -l
}

nbrVisits=$(compute_visits $1)
nbrActions=$(compute_actions $1)

curl \
    -X POST \
    -H "Authorization: Bearer $TOKEN" \
    -H "Content-Type: application/json" \
    -d "{\"visits\": $nbrVisits, \"actions\": $nbrActions}" \
    ${BASE_URL}/api/v1/kpis
```

mzur commented on tests/Http/Controllers/Views/KpiControllerTest.php:
```diff
@@ -0,0 +1,42 @@
+        $response = $this->get($kpiPath);
+        $response->assertStatus(200);
+
+        // Indices > 5 are set to 5 (last month index)
```
Isn't it 6 in the controller? Maybe you can perform a redirect in that case or a 404 not found? This can be tested more easily, too.

lehecht committed:
Make queries more efficient

lehecht committed:
Merge kpi tables

lehecht committed:
Use only rules to validate request

lehecht committed:
Add token prefix

lehecht committed:
Update test

lehecht committed:
Read values from request input

lehecht committed:
Simplify lines

lehecht committed:
Add redirect for invalid month indices

lehecht committed:
Fix wrong namespace

lehecht committed:
Define middleware class directly

lehecht committed:
Simplify lines

lehecht committed:
Fix get methods

Get should return sum of requests of a month.

lehecht committed:
Simplify query

lehecht committed:
Add missing imports

lehecht committed:
Add comment

lehecht committed:
Make index necessary

lehecht committed:
Add inital index to route

lehecht committed:
Remove unused index

lehecht committed:
Add apiDoc

lehecht committed:
Add tests

mzur commented:
If this is ready for review again, please click the button :wink: You can ignore the Bash script for now, we can discuss that in the thread above.

lehecht committed:
Update script

Count only get requests to homepage.
Enable reading of zipped files by using zgrep.

lehecht commented:
@mzur 
I pushed the updated script. Now I have revised everything.

mzur commented on src/Http/Controllers/RequestController.php:
```diff
@@ -9,15 +9,27 @@
+     *       "visits": 10,
+     *       "actions": 20
+     *    }
+     * }
```
```suggestion
     * {
     *     "visits": 10,
     *     "actions": 20
     * }
```

mzur commented on src/Http/Controllers/RequestController.php:
```diff
@@ -9,15 +9,27 @@
-     * @param StoreRequest $request containing the json object
+     * @api {post} /kpis
+     * @apiPermission token-verifyed
+     * @apiDescription Saves actions/visits of a serverlog file
```
```suggestion
     * @apiPermission kpi-token
     * @apiDescription Save actions/visits KPIs
     * @apiParam {Number} visits The number of GET requests to `/` (without bots).
     * @apiParam {Number} actions The number of PUT/POST/DELETE requests (without bots and haertbeat).
```

The `apiPermission` value needs to be defined somewhere. You can do this by pasting this block above the controller method:

```php
/**
 * @apiDefine kpi-tpken KPI Token
 * The request must provide the token configured in the KPI module.
 */
```

mzur commented on src/Http/Controllers/RequestController.php:
```diff
@@ -9,15 +9,27 @@
-        Requests::save($res['actions'], $res['visits']);
-        return;
+        $visits = $visits = $request->input('visits');
+        $actions = $visits = $request->input('actions');
```
This (visits=actions) should have been caught by the tests. please update the tests so they check the actual values in the database..

```suggestion
        $visits = $request->input('visits');
        $actions = $request->input('actions');
```

mzur commented on src/Http/Requests/StoreRequest.php:
```diff
@@ -15,38 +15,8 @@ class StoreRequest extends FormRequest
-        });
+            'actions' => 'required|integer',
+            'visits' => 'required|integer'
+            ];
```
```suggestion
        ];
```

mzur commented on tests/Http/Controllers/RequestControllerTest.php:
```diff
@@ -21,34 +21,50 @@ public function setUp(): void
+            $this->route,
+            ['visits' => '94556', 'actions' => '21602']
+        )->assertStatus(200);
     }
```
Please add a check for the actual values stored in the DB (see comment above).

mzur commented on src/public/assets/scripts/main.js:
Please run `npm run prod` to make this file disappear again.

lehecht replied:
Should this file be deleted or just be empty?

mzur replied:
Deleted.

mzur commented on src/Http/Controllers/Views/KpiController.php:
```diff
@@ -18,8 +18,9 @@ class KpiController extends Controller
+            abort(404);
+        }
 
         $date = Carbon::now()->subMonths(6 - $idx);
```
The month selection doesn't work quite yet. Today is April and if I select "Feb", I'm redirected to `admin/kpis/5`. This gives `subMonths(6 - 5)` which returns March instead of February.

When I visit the page in April I want to see the KPIs of March (`subMonths(1)`) and for February it should call `subMonths(2)`.

mzur commented on src/resources/views/show.blade.php:
Please use `number_format()` (with default parameters) to display the numbers. It makes large numbers easier to read.

lehecht committed:
Use number_format method

lehecht committed:
Change indention

lehecht committed:
Fix wrong month index calculation

lehecht committed:
Fix copy paste errors

lehecht committed:
Add tests to check stored values

lehecht committed:
Commit npm run prod

lehecht committed:
Remove unused import

lehecht committed:
Extend api doc

lehecht committed:
Delete public/assets/scripts/main.js

lehecht committed:
Change indent in api doc

lehecht committed:
Rename variables/methods for better understanding

mzur reviewed (approved):
Thanks! I'll take it from here.
